### PR TITLE
Implement XEP-0157: Contact Addresses for XMPP Services

### PR DESCRIPTION
**Pull request by lissine0**
yax.im
conversations.im

### CODE REVIEW

File: Monal/Classes/MLIQProcessor.m [hunk 1]
```diff
@@ -548,6 +548,25 @@ +(BOOL) processRosterWithAccount:(xmpp*) account andIqNode:(XMPPIQ*) iqNode
...
+        {
+            NSMutableArray* addresses = [NSMutableArray array];
+            for (MLXMLNode* valueElement in field.children)
+            {
+                [addresses addObject:valueElement.data];
+            }
+            resultDictionary[fieldName] = addresses;
+        }
+    }
+    account.connectionProperties.serverContactAddresses = resultDictionary;
```
Comment (lissine0) [inline Monal/Classes/MLIQProcessor.m]: Is this fine? or is [ copy]  /  [ mutableCopy] more appropriate?

File: Monal/Classes/MLIQProcessor.m [hunk 2]
```diff
@@ -548,6 +548,25 @@ +(BOOL) processRosterWithAccount:(xmpp*) account andIqNode:(XMPPIQ*) iqNode
     //query external services to learn stun/turn servers
     if([features containsObject:@"urn:xmpp:extdisco:2"])
         [account queryExternalServicesOn:iqNode.fromUser];
+
+    //get the server's contact addresses (XEP-0157)
```
Comment (lissine0) [inline Monal/Classes/MLIQProcessor.m:552]: I considered moving this section to a separate handler, but I think there is no need to send two separate IQs that are identical. See [xmpp.m L3959-L3961](https://github.com/monal-im/Monal/blob/develop/Monal/Classes/xmpp.m#L3959-L3961)

File: Monal/Classes/MLIQProcessor.m [hunk 3]
```diff
@@ -548,6 +548,25 @@ +(BOOL) processRosterWithAccount:(xmpp*) account andIqNode:(XMPPIQ*) iqNode
...
+    XMPPDataForm* dataForm = [iqNode findFirst:@"{http://jabber.org/protocol/disco#info}query/\\{http://jabber.org/network/serverinfo}result\\"];
+    NSMutableDictionary* resultDictionary = [NSMutableDictionary dictionary];
+    NSString* fieldName;
```
Comment (lissine0) [inline Monal/Classes/MLIQProcessor.m]: Side question: would it be more efficient if NSMutableString is used here?

File: Monal/Classes/MLIQProcessor.m [hunk 4]
```diff
@@ -548,6 +548,25 @@ +(BOOL) processRosterWithAccount:(xmpp*) account andIqNode:(XMPPIQ*) iqNode
...
+    for (MLXMLNode* field in dataForm.children)
```
Comment (tmolitor-stud-tu) [inline Monal/Classes/MLIQProcessor.m]: Read the data-forms part of my xml query article and check out the dictionary and array interfaces in XMPPDataForm.h as well.

File: Monal/Classes/MLIQProcessor.m (hunk 1 above)
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m]: having this non-mutable would be better...

File: Monal/Classes/MLIQProcessor.m (hunk 3 above)
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m]: use NSMutableString where?

File: Monal/Classes/MLIQProcessor.m (hunk 2 above)
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m:552]: no, sending two iqs would be just stupid, your solution to add it here is great :)

**Review by tmolitor-stud-tu — CHANGES_REQUESTED** [review]

File: Monal/Classes/MLIQProcessor.m (hunk 3 above)
Comment (lissine0) [reply Monal/Classes/MLIQProcessor.m]: `NSMutableString* fieldName;` instead of `NSString* fieldName;`

File: Monal/Classes/MLIQProcessor.m (hunk 1 above)
Comment (lissine0) [reply Monal/Classes/MLIQProcessor.m]: So do you mean just
`account.connectionProperties.serverContactAddresses = [resultDictionary copy];`

Or do you also mean changing `MLXMPPConnection.serverContactAddresses`'s type from `NSMutableDictionary*` to `NSDictionary*`

Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m]: both, copy and change to NSDictionary

File: Monal/Classes/MLIQProcessor.m (hunk 3 above)
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m]: but the field name isn't mutated at all, why use NSMutableString? You'd even have to mutably copy the NSString over into an NSMutableString as opposed to just using the NSString directly as is.

File: Monal/Classes/MLIQProcessor.m (hunk 4 above)
Comment (tmolitor-stud-tu) [reply Monal/Classes/MLIQProcessor.m]: This still needs to be addressed...

**Comment by tmolitor-stud-tu** [comment]
quick heads up: this is still missing the dataforms rewrite to be mergeable...

**Comment by lissine0** [comment]
Just as an FYI:

- the `%fieldName` extraction command doesn't work: it doesn't have any effect.
For example, `[iqNode findFirst:@"{http://jabber.org/protocol/disco#info}query/\\{http://jabber.org/network/serverinfo}result%%admin-addresses\\"];` returns the whole dataForm, as if `%%admin-addresses` wasn't there.
On the other hand, XMPPDataForm's `getField:(NSString*) name` works; and the `@fieldName` extraction command works.

 And

- XMPPDataForm's `objectAtIndexedSubscript:(NSInteger) idx` doesn't work.
In fact trying `[dataForm objectAtIndexedSubscript:0];` causes an exception: `Exception  NSException *   "*** -[__NSArray0 objectAtIndex:]: index 0 beyond bounds for empty array"`, eventhough the server does have published contact addresses.
XMPPDataForm.h does have a [comment](https://github.com/monal-im/Monal/blob/8f0a624/Monal/Classes/XMPPDataForm.h#L38) above the method signature, saying the interface is incomplete, so I assume this is expected.

**Comment by tmolitor-stud-tu** [comment]
Why did you double the `%`? You were qurying for the field `%admin-addresses` which indeed isn't there.

But why use the `%` operator at all? That will return a dictionary with additional information, but you'll only need the field value, so the `@` operator should be enough...

**Comment by tmolitor-stud-tu** [comment]
`objectAtIndexedSubscript` will only work for data-forms having multiple item elements encapsulating the form fields, see [example 8 of XEP-0004](https://xmpp.org/extensions/xep-0004.html#example-8).

**Comment by tmolitor-stud-tu** [comment]
No, the "incomplete interface" comment means that it does not implement everything an `NSArray` or `NSDictionary` is capable of.

**Comment by lissine0** [comment]
> Why did you double the `%`? You were qurying for the field `%admin-addresses` which indeed isn't there.

%% is needed to escape the % inside the string, otherwise it'll be looking for a conversion format (%@, %d etc.)
In fact, including a single % character results in a compilation error.

> But why use the `%` operator at all?

I don't need it in this case. I was just trying different things.

**Comment by tmolitor-stud-tu** [comment]
ah, this is a interference between the `NS_FORMAT_FUNCTION` annotation (triggering compiler errors for incorrectly used format specifiers) of the `find:` method and the query implementation not allowing a format specifier at that position but using it as an "ordinary" extraction operator...

I'll fix that and use another operator instead of `%`, but the `@` should still work just fine.

**Comment by lissine0** [comment]
I tested eb680c6f1 and the `&` extraction command works correctly

**Pull request merged**